File: 프로그래머스/3/42898. 등굣길/등굣길.py
from collections import deque
def solution(m, n, puddles):
    dp=[[0]*m for _ in range(n)]
    for x,y in puddles:
        dp[y-1][x-1]=-1

    # dp[x][y]=dp[x-1][y] + dp[x][y-1]
    
    # print(dp)
    dp[0][0]=1
    for i in range(n):
        for j in range(m):
            if dp[i][j]==-1:
                dp[i][j]=0
                continue
            if i>0 :
                dp[i][j]+=dp[i-1][j]
            if j>0 :
                dp[i][j]+=dp[i][j-1]
            
    # print(dp)

    return dp[n-1][m-1]%1000000007

File: 프로그래머스/3/42898. 등굣길/등굣길.java
import java.util.*;
class Solution {
    public int solution(int m, int n, int[][] puddles) {
        int mod = 1000000007;
        
        // 맵 초기화
        int[][] dp = new int[n+1][m+1];
        
        // 장애물 map 추가
        for (int[] puddle:puddles){
            dp[puddle[1]][puddle[0]] = -1;
        }
        
        dp[1][1] = 1;
        
        // dp[r][c] = dp[r-1][c] + dp[r][c-1]
        for (int r=1;r<=n;r++){
            for (int c=1;c<=m;c++){
                if (dp[r][c]==-1){
                    continue;
                }
                if (r-1>0 && dp[r-1][c]!=-1){
                    dp[r][c]+=dp[r-1][c];
                }
                if (c-1>0 && dp[r][c-1]!=-1){
                    dp[r][c]+=dp[r][c-1];
                }
                dp[r][c] %= mod;
            }
        }
        
        return dp[n][m]%mod;
    }
}